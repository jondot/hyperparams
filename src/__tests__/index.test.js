const { extract } = require('../../src/index')

const body = {
  user: { name: 'joe', age: 31, _id: 123 },
  token: 2,
  posts: {
    total: 2,
    meta: { agent: 'superagent' },
    items: [{ _id: 1, title: 'foobar' }, { _id: 2, title: 'goo!' }],
  },
}

it('flat props', () =>
  expect(extract('user', ['name', 'age'])(body)).toMatchSnapshot())
it('nested w/array', () =>
  expect(
    extract('posts', ['total', { items: ['title'] }])(body)
  ).toMatchSnapshot())
it('nested w/object', () =>
  expect(
    extract('posts', [{ meta: ['agent'] }, 'total'])(body)
  ).toMatchSnapshot())
it('flat w/object', () =>
  expect(extract('posts', ['total', 'meta'])(body)).toMatchSnapshot())
it('object but w/missing props', () =>
  expect(() => extract('posts', ['total', 'meta', 'foobar'])(body)).toThrow())
it('missing all throws', () =>
  expect(() =>
    extract('posts', ['foo', 'bar', { baz: ['poop'] }])(body)
  ).toThrow())
it('missing but merged', () =>
  expect(
    extract('posts', ['foo', 'bar', { baz: ['poop'] }], {
      foo: 1,
      bar: 2,
      baz: { poop: 3 },
    })(body)
  ).toMatchSnapshot())
it('missing root throws', () =>
  expect(() => extract('foobar', ['baz'])(body)).toThrow())
it('no props given means everything', () =>
  expect(extract('posts', [])(body)).toMatchSnapshot())
it('no props for nested array means verbatim array items', () =>
  expect(extract('posts', [{ items: [] }])(body)).toMatchSnapshot())
it('empty string root means shallow root object', () =>
  expect(extract('', ['token'])(body)).toMatchSnapshot())
/* 
 *
 *  user_params = body => from(body).extract('user').permit('name', {defaults}).value
 *  user_params = extract('user').permit('name', {defaults})
 *
 *
 */
