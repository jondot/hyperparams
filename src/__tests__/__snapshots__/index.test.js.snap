// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaults should not override existing 1`] = `
Object {
  "total": 2,
}
`;

exports[`empty string root means shallow root object 1`] = `
Object {
  "token": 2,
}
`;

exports[`flat props 1`] = `
Object {
  "age": 31,
  "name": "joe",
}
`;

exports[`flat w/object 1`] = `
Object {
  "meta": Object {
    "agent": "superagent",
  },
  "total": 2,
}
`;

exports[`missing but merged 1`] = `
Object {
  "bar": 2,
  "baz": Object {
    "poop": 3,
  },
  "foo": 1,
}
`;

exports[`nested w/array 1`] = `
Object {
  "items": Array [
    Object {
      "title": "foobar",
    },
    Object {
      "title": "goo!",
    },
  ],
  "total": 2,
}
`;

exports[`nested w/object 1`] = `
Object {
  "meta": Object {
    "agent": "superagent",
  },
  "total": 2,
}
`;

exports[`no props for nested array means verbatim array items 1`] = `
Object {
  "items": Array [
    Object {
      "_id": 1,
      "title": "foobar",
    },
    Object {
      "_id": 2,
      "title": "goo!",
    },
  ],
}
`;

exports[`no props given means everything 1`] = `
Object {
  "items": Array [
    Object {
      "_id": 1,
      "title": "foobar",
    },
    Object {
      "_id": 2,
      "title": "goo!",
    },
  ],
  "meta": Object {
    "agent": "superagent",
  },
  "total": 2,
}
`;
